### PR TITLE
[lodash] Add a type predicate to `isObjectLike` and explain why one cannot be added to `isPlainObject`

### PR DESCRIPTION
**Pull request by isti115**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

If changing an existing definition:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [`isObjectLike`](https://github.com/lodash/lodash/blob/main/src/isObjectLike.ts), [`isPlainObject`](https://github.com/lodash/lodash/blob/main/src/isPlainObject.ts)
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

### CODE REVIEW

**Review by aj-r — CHANGES_REQUESTED** [review]
This change has been proposed several times before, but unfortunately it doesn't work because it breaks the negative version of the type guard (e.g. the type in the `else` block). AFAIK there's no way to define a reasonable type guard for these functions, so unfortunately we're stuck with `boolean`.

**Comment by aj-r** [comment]
Yeah that's a good idea

File: types/lodash/common/lang.d.ts [hunk 1]
```diff
@@ -1098,6 +1098,10 @@ declare module "../index" {
          * // => false
          */
         isObjectLike(value?: any): boolean;
+
+        // NOTE: The seemingly logical `value is object` type predicate cannot
+        // be added, as it would break the negative version of the type guard.
+        // https://github.com/DefinitelyTyped/DefinitelyTyped/pull/69915#pullrequestreview-2143148704
```
Comment (aj-r) [inline types/lodash/common/lang.d.ts:1104]: I explained in a little more detail here: https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/68867#discussioncomment-8671502. In fact it looks like my past self corrects me... The type guard should work fine on isObjectLike, just not isPlainObject. Sorry for the confusion!

**Review by aj-r — CHANGES_REQUESTED** [review]

File: types/lodash/common/lang.d.ts (hunk 1 above)
Comment (isti115) [reply types/lodash/common/lang.d.ts]: Sorry for the long delay, I have finally got to wrapping this up. In the latest commit I add the type guard to `isObjectLike` and the explanation to `isPlainObject`.

File: types/lodash/lodash-tests.ts [hunk 2]
```diff
@@ -4495,6 +4495,12 @@ fp.now(); // $ExpectType number
     _(42).isObjectLike(); // $ExpectType boolean
     _.chain([]).isObjectLike(); // $ExpectType PrimitiveChain<boolean>
     fp.isObjectLike(anything); // $ExpectType boolean
+    if (fp.isObjectLike(anything)) {
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts:4498]: I think this test is failing. You'll probably need to manually update the `fp` version (in `fp.d.ts` IIRC).

**Review by aj-r — CHANGES_REQUESTED** [review]
Thanks for fixing this!

**Comment by isti115** [comment]
@aj-r Sorry for the drawn out resolution of this PR!

I've tried running the `generate-fp.ts` script as requested by the header of `fp.d.ts`:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d5a5c3b0ef50b7277750ed631c3d640b27272143/types/lodash/fp.d.ts#L1-L3

But it didn't manage to execute using the provided commands, so I used `tsx` instead, but even after it succeeded, it also changed a bunch of other things in the file, so I finally just resorted to editing the relevant part manually.

Also, some tests are still failing for me, but they don't seem to be related to my changes.

File: types/lodash/common/lang.d.ts [hunk 3]
```diff
@@ -1097,13 +1097,13 @@ declare module "../index" {
          * _.isObjectLike(null);
          * // => false
          */
-        isObjectLike(value?: any): boolean;
```
Comment (gabritto) [inline types/lodash/common/lang.d.ts:1100]: This doesn't look correct: type `object` includes functions in TypeScript, but `isObjectLike` returns `false` for functions. For this to work, you'd have to use a type that expresses all non-function objects. Maybe using `Record`?

**Review by gabritto — CHANGES_REQUESTED** [review]

File: types/lodash/common/lang.d.ts (hunk 3 above)
Comment (gabritto) [reply types/lodash/common/lang.d.ts:1100]: On second thought, I don't think TS can actually express the type of non-function objects at the moment, actually.

**Review by Sec32fun32 — APPROVED** [review]

**Pull request closed without merging**